### ISSUE TITLE
Umbraco 8 Support

### ISSUE THREAD

**Issue by TomSteer** [JimBobSquarePants/UmbMapper#8]
Hey James, I've used this library on a few of our projects now and found it great to work with so high five for creating such a great package. I was just wondering if Umbraco 8 was something you were considering adding support for? Cheers

**Comment by TomSteer** [comment]
Ironically issue number 8 :laughing:

**Comment by JimBobSquarePants** [comment]
@TomSteer It's actually a big job and I would need help doing so. 

Converting the codebase itself should be a fairly trivial operation. (Though we'd have to convert the core API  +  factory only, no other extensions for now) Converting the tests however looks to me to be a massive undertaking.

**Comment by euanrae** [comment]
hi guys,

I've forked the repo and made some minor changes and looks like I've got it working with v8.

One thing that I haven't worked out yet is to create an UmbracoContext if one doesn't exist, so I'm not sure if this will work with an API controller.

I'm going through to and adding all the tests; it's a bit tricky with the change of Composers, UmbracoContext, etc.

Will submit a PR once all the tests are done, though should probably add some more to make sure the it still works when using an API controller.

**Comment by JimBobSquarePants** [comment]
@euanrae Nice!!!

Thanks absolutely brilliant. I really wasn't looking forward to updating the tests. Thanks!

**Comment by euanrae** [comment]
Happy to help @JimBobSquarePants ! I have a few big projects coming up in 8 and really want to use this :)

So, this pretty much works, I'm planning on setting up a test site to test most of the different entities to be mapped, and I think there's some things that may need to be looked at first.

TLDR
The main question is: 
- does this package need to be able to have tests written against it (which some of the changes in Umbraco 8 have made it quite difficult), which may require some larger changes, AND a sample site, OR,
-  given that it works, is a test site with most of the mapping types in there and working good enough.

I'd prefer to change it to have full test coverage, but (I think) the change in V8 may require some bigger changes to get everything to work & be testable.

I've been staring at this for too long and gone down the rabbit hole of trying to mock all the umbraco 8 stuff, so I'm going to take a break for a few days.  I'm happy to help where I can (I love this package and really want to use it soon in v8!!), and I don't want to go completely off the rails and restructuring a lot of stuff if there's something I'm missing, or over-complicating it - let me know what your thoughts are re approach

//TLDR

Long Version
Overlapping, but slightly separate issues:
1. UmbracoContext - this is treated differently in V8, so you can't just set the current context (or use UmbracoContext.Current), e.g
```
            UmbracoContext.EnsureContext(
                dummyHttpContext,
                ApplicationContext.Current,
                new WebSecurity(dummyHttpContext, ApplicationContext.Current),
                UmbracoConfig.For.UmbracoSettings(),
                UrlProviderResolver.Current.Providers,
                false);
```
you need to have an instance of IUmbracoContextFactory, which can be injected in.  I've done a quick PoC to create an UmbMapper service that takes in (IUmbracoContextFactory) as a ctor parameter, and that does work correctly, but it would require a) that IUmbracoContextFactory instance to be passed to all the mappers which isn't very nice or b) a bigger structural change.
The other issue with this approach is that
2. UmbracoHelpers (relates to the point above)
```
public MembershipHelper Members => new MembershipHelper(this.UmbracoContext); 
public UmbracoHelper Umbraco => new UmbracoHelper(this.UmbracoContext);
```
isn't really an option any more, the other options are to pass in the helpers as parameters somewhere, or do this:
```
public UmbracoHelper Umbraco => global::Umbraco.Web.Composing.Current.UmbracoHelper;
public MembershipHelper Members => this.Umbraco.MembershipHelper;
```
which isn't great either.
3. Members are handled slightly differently than content and media.  Using the IUmbracoContextFactory, you can get an instance of IPublishedContentCache and IPublishedMediaCache, but with members, if you have a parameter if type (IPublishedMemberCache) in your controller ctor, it will get injected in.
4. Testing - a lot of the Umbraco constructors are now internal, and the difference in the way IPublishedContentCache, IPublishedMediaCache and IPublishedMemberCache work means it could be a bit messy to get instances of these things into UmbMapper to use for mapping.

**Comment by euanrae** [comment]
(Sorry for the spam, but...)
I've looked into the Umbraco source and realised mocking a property value editor seems to be the way to go, so I've done that which has helped pass all the tests.

I need to tidy up the tests a bit more, and still need to work out how this will work out without being able to call UmbracoContext.EnsureContext(...) for Apis.

**Comment by JimBobSquarePants** [comment]
@euanrae I'm sure I saw a message in my emails (I thought from here) where you were asking questions about unit tests. Have I gone mad or does that exist?

**Comment by euanrae** [comment]
@JimBobSquarePants - yes, there was a question but I ended up deleting it, as I ended up working out the answer after digging into the Umbraco source a bit more.  Basically, Mocking UmbracContext is now more difficult/messy as the constructor is internal, but I managed to work out a way to do it.
Am in the process of creating a test project now to double check everything works as expected.

**Comment by JimBobSquarePants** [comment]
@euanrae Great to hear! Thanks again!

**Comment by euanrae** [comment]
@JimBobSquarePants - happy to help.

I'm currently getting the following exception:System.EntryPointNotFoundException: Entry point was not found.

When the view is trying to access the the lazy loaded slides:
@if (Model.Content.Gallery.Any())

Any clues what this might be?

The first thing that comes to mind is the new 'Composing' stuff in Umbraco 8 ( I have the mappers being set in an IComponent.Initialize method )

**Comment by JimBobSquarePants** [comment]
I’ll have a look....

Please bear in mind that one of the reasons I chose to not make the effort to migrate to v8 was that I considered it “not ready”. Despite the protestations of hq I believe it is, at best, beta software. 

If errors occur they are likely due to the incomplete infrastructure of v8.

**Comment by euanrae** [comment]
Fair enough; I had a feeling that may be the case.  I'm hoping it was an error on my part, and that as you have a better understanding of the codebase it would be more obvious to you.

Thanks

**Comment by JimBobSquarePants** [comment]
@euanrae If you create a draft PR against the develop branch and make sure I have write access (I think that's default actually but best to check) then I can help you get the last bits up and running.

**Comment by euanrae** [comment]
@JimBobSquarePants - thanks. Draft PR has been submitted

**Comment by euanrae** [comment]
It looks like using Umbraco 8 tests is causing issues in the build - I am working on a way to mock umbraco context without relying on v8 umbraco.tests assembly

**Comment by JimBobSquarePants** [comment]
Had a quick look at the PR. The library needs refactoring to use DI. Then the context factory can be injected. A fair amount of work.

**Comment by euanrae** [comment]
Yeah, accessing it through Composting.Current didn't seem ideal.  Will abandon the PR and look at refactoring.

I assume this means
- changing UmbMapperRegistry to be an instance instead of static + injected as a singleton
- changing the map to methods to be in some type of service instead of extension methods
- have IUmbracoContextFactory as a parameter where needed

I assume there will also be some restructuring required but will need to look into it further to work out where/how.

**Comment by JimBobSquarePants** [comment]
@euanrae Yeah, all you've said seems sensible. 

On the plus side though we don't need to recreate any of the [object extensions](https://github.com/JimBobSquarePants/UmbMapper/blob/4e2a8eff6836b498f7a542120e63d35b9911dafa/src/UmbMapper/Extensions/ObjectExtensions.cs) code as that was a hangover from poor performing V7 versions (long since updated), nor include any of the libraries other than the code UmbMapper library since I don't think V8 supports any of them.

If you create a WIP PR I'll jump in and help out when I can.

**Comment by euanrae** [comment]
thanks, sounds good

**Comment by nhaberl** [comment]
Just asking if you are still working on it ... would be great to have a tiny, fast mapper in the toolset also for v8

**Comment by euanrae** [comment]
Have been working on this on / off when I can, it still needs more work though:

https://github.com/JimBobSquarePants/UmbMapper/pull/10
